File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch {
        case card == "ace":
    		return 11
        case card == "king" || card == "queen" || card == "jack" || card == "ten":
    		return 10
        case card == "nine":
    		return 9
        case card == "eight":
    		return 8
        case card == "seven":
    		return 7
        case card == "six":
    		return 6
        case card == "five":
    		return 5
        case card == "four":
    		return 4
        case card == "three":
    		return 3
        case card == "two":
    		return 2
        default:
    		return 0
    }
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
    cardSum := ParseCard(card1) + ParseCard(card2)
	switch {
        case card1 == "ace" && card2 == "ace":
    		return "P"
        case cardSum == 21:
    		if (ParseCard(dealerCard) >= 10) {
                return "S"
            } else {
            	return "W"
            }
    	case cardSum <= 20 && cardSum >= 17:
    		return "S"
        case cardSum <= 16 && cardSum >= 12:
    		if (ParseCard(dealerCard) >=7 ) {
                return "H"
            } else {
            	return "S"
            }
    	default:
    		return "H"
    }
}

File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when "ace" then 11
    when "king" then 10
    when "queen" then 10
    when "jack" then 10
    when "ten" then 10
    when "nine" then 9
    when "eight" then 8
    when "seven" then 7
    when "six" then 6
    when "five" then 5
    when "four" then 4
    when "three" then 3
    when "two" then 2
    else 0
    end
  end

  def self.card_range(card1, card2)
    card1_score = parse_card(card1)
    card2_score = parse_card(card2)
    total_score = card1_score + card2_score

    case total_score
    when (4..11) then "low"
    when (12..16) then "mid"
    when (17..20) then "high"
    else "blackjack"
    end  
  end

  def self.first_turn(card1, card2, dealer_card)
    return "P" if card1 == "ace" && card2 == "ace"

    dealer_score = parse_card(dealer_card)

    case card_range(card1, card2)
    when "blackjack"
      if dealer_score < 10
        "W"
      else 
        "S"
      end
    when "high"
      "S"
    when "mid"
      if dealer_score >= 7
        "H"
      else 
        "S"
      end
    else "H"
    end
  end
end
